private/buf/bufsync/error.go

```diff
@@ -0,0 +1,114 @@
+	// ReadModuleErrorCodeBuildModule happens when the read module errors building.
+	ReadModuleErrorCodeBuildModule
+	// ReadModuleErrorCodeUnexpectedName happens when the read module has a different name than
+	// expected, usually the one in the branch HEAD commit.
+	ReadModuleErrorCodeUnexpectedName
+)
+
+// ReadModuleError is an error that happens when trying to read a module from a module directory in
+// a git commit.
+type ReadModuleError struct {
+	err       error
+	code      ReadModuleErrorCode
+	branch    string
+	commit    string
+	moduleDir string
+}
+
+func (e *ReadModuleError) Error() string {
+	return fmt.Sprintf(
+		"read module in branch %s, commit %s, directory %s: %s",
+		e.branch, e.commit, e.moduleDir, e.err.Error(),
+	)
+}
+
+// ErrorHandler handles errors reported by the Syncer before or during the sync process.
+type ErrorHandler interface {
+	// StopLookback is invoked when deciding on a git start sync point.
+	//
+	// For each branch to be synced, the Syncer travels back from HEAD looking for modules in the
+	// given module directories, until finding a commit that is already synced to the BSR, or the
+	// beginning of the Git repository.
+	//
+	// The syncer might find errors trying to read a module in that directory. Those errors are sent
+	// to this function to decide if the Syncer should stop looking back or not, and choose the
+	// previous one (if any) as the start sync point.
+	//
+	// e.g.: The git commits in topological order are: `a -> ... -> z (HEAD)`, and the modules on a
+	// given module directory are:
+	//
+	// commit | module name or failure | could be synced? | why?
+	// ----------------------------------------------------------------------------------------
+	// z      | buf.build/acme/foo     | Y                | HEAD
+	// y      | buf.build/acme/foo     | Y                | same as HEAD
+	// x      | buf.build/acme/bar     | N                | different than HEAD
+	// w      | unnamed module         | N                | no module name
+	// v      | unbuildable module     | N                | module does not build
+	// u      | module not found       | N                | no module name, no 'buf.yaml' file
+	// t      | buf.build/acme/foo     | Y                | same as HEAD
+	// s      | buf.build/acme/foo     | Y                | same as HEAD
+	// r      | buf.build/acme/foo     | N                | already synced to the BSR
+	//
+	// If this func returns 'true' for any `ReadModuleErrorCode`, then the syncer will stop looking
+	// when reaching the commit `r` because it already exists in the BSR, select `s` as the start sync
+	// point, and the synced commits into the BSR will be [s, t, x, y, z].
+	//
+	// On the other hand, if this func returns true for `ReadModuleErrorCodeModuleNotFound`, the
+	// syncer will stop looking when reaching the commit `u`, will select `v` as the start sync point,
+	// and the synced commits into the BSR will be [x, y, z].
+	StopLookback(err *ReadModuleError) bool
```

Comment:
This way, if we want in the future, a flag or config could control in which scenarios we stop looking back, and which ones we skip.